packages/client/src/pages/common/nav/create-profile-dialog.tsx

```diff
@@ -62,31 +79,70 @@ const StyledContainer = styled.div`
 // TODO: form validation, Firebase hookup
 const CreateProfileDialog = ({ open, onClose }: CreateDialogProps) => {
   const classes = useStyles();
-  const [values, setValues] = React.useState<ProfileState>(DEFAULT_FIELDS);
+  const [values, setValues] = useState<ProfileState>(DEFAULT_FIELDS);
+  const [errors, setErrors] = useState<ProfileErrors>(DEFAULT_ERRORS);
   const dispatch = useAppDispatch();
 
   const handleClose = () => {
     setValues(DEFAULT_FIELDS);
     onClose();
   };
 
-  const handleSignUp = () => {
-    dispatch(
-      signup({
-        username: values.username,
-        password: values.password,
-        email: values.email,
-      }),
-    )
-      .then(unwrapResult)
-      .then((data) => alert("Signup succeeded!"))
-      .catch((error) => alert("Signup failed.:" + error));
-    handleClose();
+  const handleSignUp = async () => {
+    try {
+      const { username, email } = values;
+      const validationResult = await userService.validate({
+        username,
+        email,
+      });
+      let isValid = true;
+      if (validationResult.emailExists) {
+        setErrors({ ...errors, email: "The email already exists" });
+        isValid = false;
+      }
+
+      if (validationResult.usernameExists) {
+        setErrors({ ...errors, username: "The username already exists" });
+        isValid = false;
+      }
+
+      if (isValid) {
+        const dispatchedAction = await dispatch(
+          signup({
+            username: values.username,
+            password: values.password,
+            email: values.email,
+          }),
+        );
+        unwrapResult(dispatchedAction);
+        handleClose();
+      }
+    } catch (err) {
+      if (err instanceof ActionableError) {
+        switch (err.errorCode) {
+          case SignupErrorCode.DUPLICATE_EMAIL:
+          case SignupErrorCode.INVALID_EMAIL:
+            setErrors({ ...errors, email: err.message });
+            break;
+          case SignupErrorCode.DUPLICATE_USERNAME:
+            setErrors({ ...errors, username: err.message });
+            break;
+          case SignupErrorCode.WEAK_PASSWORD:
+            setErrors({ ...errors, password: err.message });
+        }
+      } else {
+        alert(
+          "There was an unexpected error while signing up, please try again",
+        );
+      }
+    }
   };
 
   const handleChange =
     (prop: keyof ProfileState) =>
     (event: React.ChangeEvent<HTMLInputElement>) => {
+      // TODO: Do validation on field
```

Comment:
For convenience we probably want to setup form validation as well rather than just waiting for server response.